File: C++/Fizzbuzz_divide_and_conquer.cpp
// Recursive divide and conquer FizzBuzz, something like a binary search
// Author: @lusmoura

#include <bits/stdc++.h>
using namespace std;

string get_string(int num) {
    string ans = "";
    if (num % 3 == 0) ans += "Fizz";
    if (num % 5 == 0) ans += "Buzz";
    if (ans.empty()) ans = to_string(num);

    return ans;
}

void get_fizzbuzz(int curr_min, int curr_max, vector<string>& fizzbuzz_list) {
    if (curr_min > curr_max) return;
    if (curr_min == curr_max) {
        fizzbuzz_list[curr_min] = get_string(curr_min);
        return;
    }

    int mid = (curr_max+curr_min)/2;
    get_fizzbuzz(mid+1, curr_max, fizzbuzz_list);
    get_fizzbuzz(curr_min, mid, fizzbuzz_list);
}

int main() {
    int n = 100;
    vector<string> fizzbuzz_list(n+1);
    get_fizzbuzz(1, n, fizzbuzz_list);
    
    for (int i = 1; i <= n; i++) {
        cout << i << ": " << fizzbuzz_list[i] << '\n';
    }

    return 0;
}

File: Java/FizzBuzz_DivideAndConquer.java
// Divide and Conquer version of FizzBuzz
// Author: @evamy

import java.util.Arrays;

public class FizzBuzz_DivideAndConquer {

    static void fizzBuzz(int start, int end, String[] outArray) {
        if (start > end) return;
        if (start == end) {
            outArray[start-1] = next(start);
            return;
        }

        int mid = (start+end) / 2;
        fizzBuzz(mid+1, end, outArray);
        fizzBuzz(start, mid, outArray);
    }

    static String next(int num) {
        String ans = "";
        if (num % 3 == 0) ans += "Fizz";
        if (num % 5 == 0) ans += "Buzz";
        if (ans.isEmpty()) ans += num;

        return ans;
    }

    public static void main(String[] args) {
        int n = 100;

        String[] outArray = new String[n];
        fizzBuzz(1, n, outArray);

        System.out.println(Arrays.toString(outArray));
    }

}
